dynamic/discovery/discovery.go

```diff
@@ -32,128 +32,150 @@ import (
+// apiResourceRegistry is the registry of resources, kinds &
+// sub resources discovered at kubernetes server
 //
-// Resources are anchored by appropriate keys for easy
-// lookup. A resource can be looked up based on kind or
-// resource name or sub-resource name
+// Resources are anchored by appropriate keys for easy lookup.
+// A resource can be looked up either by its kind, resource name
+// or sub-resource name.
 type apiResourceRegistry struct {
-	resources, kinds, subresources map[string]*APIResource
+	resources    map[string]*APIResource
+	kinds        map[string]*APIResource
+	subresources map[string]*APIResource
 }
 
-// APIResourceManager helps discovering Kubernetes server
-// supported API groups, versions & resources.
-type APIResourceManager struct {
-	mutex sync.RWMutex
+// APIResourceDiscovery discovers kubernetes server supported
+// API groups, versions & resources
+type APIResourceDiscovery struct {
+	// DiscoveryClient to discover API resource
+	DiscoveryClient discovery.DiscoveryInterface
 
-	// discovered resources that are anchored by apiVersion
+	// GetForAPIVersionResourceFn is a functional type to get
+	// discovered API resource from provided apiVersion & resource
+	// name
 	//
 	// NOTE:
-	//	This map of registry works well, since the registry
-	// itself is a map of resources anchored by resource name.
-	// Hence, this combination of apiVersion and resource name
-	// is sufficient to find a particular API resource.
-	resources map[string]apiResourceRegistry
+	//	This can be used to mock GetForAPIVersionResource during
+	// unit test
+	GetForAPIVersionResourceFn func(apiVersion, resource string) *APIResource
 
-	// Client to discover API resource
-	Client discovery.DiscoveryInterface
+	mutex sync.RWMutex
 
-	// GetByResourceFn is a functional type to get API resource
-	// based out of apiVersion & resource name
+	// discovered resources anchored by **apiVersion**
 	//
 	// NOTE:
-	//	This can be used to mock GetByResource during unit test
-	GetByResourceFn func(apiVersion string, resource string) *APIResource
+	//	This map of registry works well, since the registry
+	// itself is a map of discovered resources anchored by
+	// resource name. Hence, this combination of apiVersion
+	// and resource name is sufficient to find a particular
+	// API resource.
+	discoveredResources map[string]apiResourceRegistry
 
 	stopCh, doneCh chan struct{}
 }
 
-// NewAPIResourceManager returns a new instance of
-// APIResourceManager
-func NewAPIResourceManager(c discovery.DiscoveryInterface) *APIResourceManager {
-	return &APIResourceManager{Client: c}
+// NewAPIResourceDiscoverer returns a new instance of APIResourceDiscoverer
+func NewAPIResourceDiscoverer(client discovery.DiscoveryInterface) *APIResourceDiscovery {
+	return &APIResourceDiscovery{
+		DiscoveryClient: client,
+	}
 }
 
-// GetByResource returns the API resource based on the provided
+// GetForAPIVersionResource returns the API resource based on the provided
```

Comment:
`GetForAPIVersionResource` does not seems very readable - while reading it at usage place I have no idea what to expect.

I can see that it returns `APIResource` for given apiVersion and resource, so I would propose something like :
* `getApiForVersionAndResource`
* `getAPIResourceFor()` - as arguments adds the context 
* `getAPIResourceForApiVersionAndResource`

actually hard to propose something as I am familiar by API of this package, just noticed that actual name is not very readable

Reply:
Great suggestion. I will go for `getAPIForVersionAndResource`